src/main/scala/rise/core/IsClosedForm.scala

```diff
@@ -6,15 +6,32 @@ import rise.core.traverse._
 import rise.core.types._
 
 object IsClosedForm {
+  case class OrderedSet[T](ordered : Seq[T], unique : Set[T])
+  object OrderedSet {
+    def one[T] : T => OrderedSet[T] = t => OrderedSet(Seq(t), Set(t))
+    def add[T] : T => OrderedSet[T] => OrderedSet[T] = t => ts =>
+      if (ts.unique.contains(t)) ts else OrderedSet(t +: ts.ordered, ts.unique + t)
+    def empty[T] : OrderedSet[T] = OrderedSet(Seq(), Set())
+    def append[T] : OrderedSet[T] => OrderedSet[T] => OrderedSet[T] = x => y => {
+      val ordered = x.ordered.filter(!y.unique.contains(_)) ++ y.ordered
```

Comment:
When writing `x append y`, should `x` not have priority on ordering? So that `[1 2 4] append [3 2] = [1 2 4 3]`? This would correspond to `x.ordered ++ y.ordered.filter(!x.unique(_)))`